sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/sources/GeneratorCheckpoint.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.nexmark.sources;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.io.UnboundedSource;
+
+/**
+ * Just enough state to be able to restore a generator back to where it was checkpointed.
+ */
+public class GeneratorCheckpoint implements UnboundedSource.CheckpointMark {
+  private static final Coder<Long> LONG_CODER = VarLongCoder.of();
+
+  /** Coder for this class. */
+  public static final Coder<GeneratorCheckpoint> CODER_INSTANCE =
+      new CustomCoder<GeneratorCheckpoint>() {
+        @Override public void encode(GeneratorCheckpoint value, OutputStream outStream)
+            throws CoderException, IOException {
+          LONG_CODER.encode(value.numEvents, outStream);
+          LONG_CODER.encode(value.wallclockBaseTime, outStream);
+        }
+
+        @Override
+        public GeneratorCheckpoint decode(InputStream inStream)
+            throws CoderException, IOException {
+          long numEvents = LONG_CODER.decode(inStream);
+          long wallclockBaseTime = LONG_CODER.decode(inStream);
+          return new GeneratorCheckpoint(numEvents, wallclockBaseTime);
+        }
+        @Override public void verifyDeterministic() throws NonDeterministicException {}
+      };
+
+  private final long numEvents;
+  private final long wallclockBaseTime;
+
+  GeneratorCheckpoint(long numEvents, long wallclockBaseTime) {
+    this.numEvents = numEvents;
+    this.wallclockBaseTime = wallclockBaseTime;
+  }
+
+  public Generator toGenerator(GeneratorConfig config) {
+    return new Generator(config, numEvents, wallclockBaseTime);
+  }
+
+  @Override
+  public void finalizeCheckpoint() throws IOException {
+    // Nothing to finalize.
+  }
+
+  @Override
+  public String toString() {
+    return String.format("Generator.GeneratorCheckpoint{numEvents:%d;wallclockBaseTime:%d}",
```

Comment:
<!--thread_id:cc_149244869_t; commit:3f46c5cd3ef9f6ffcf0708b68cc40426350241ff; resolved:1-->
<!--section:context-quote-->
> **kennknowles** wrote:
> In case you mess around here more, a nice standard way to do this is `new MoreObjects.toStringHelper(this).add("numEvents", numEvents).add("wallclockBaseTime", wallclockBaseTime).toString()`
> 
> Don't need to do it in a refactor PR, just throwing out thoughts.

<!--section:body-->
Agree. Addressed in the fixup commit